File: test/tests376.c

// Test program for ed376 signature
// gcc -O2 tests376.c ed376.c edwards376.c hash.c -o tests376

#include <stdio.h>
#include "tlsecc.h"

static int char2int(char input)
{
    if ((input >= '0') && (input <= '9'))
        return input - '0';
    if ((input >= 'A') && (input <= 'F'))
        return input - 'A' + 10;
    if ((input >= 'a') && (input <= 'f'))
        return input - 'a' + 10;
    return 0;
}

// Convert from a hex string to byte array 
static void fromHex(int ilen, const char *src, char *dst)
{
    int i,lz,len=0;
    char pad[256];
    while (src[len]!=0) len++;
    lz=2*ilen-len;
    if (lz<0) lz=0;
    for (i=0;i<lz;i++) pad[i]='0';  // pad with leading zeros
    for (i=lz;i<2*ilen;i++) pad[i]=src[i-lz];

    for (i=0;i<ilen;i++)
    {
        dst[i] = (char2int(pad[2*i]) * 16) + char2int(pad[2*i + 1]);
    }
}

static void byte2hex(char *ptr,unsigned char ch)
{
    int t=ch/16;
    int b=ch%16;
    if (t<10)
    	ptr[0]='0'+t;
    else
    	ptr[0]='a'+(t-10);
    if (b<10)
    	ptr[1]='0'+b;
    else
    	ptr[1]='a'+(b-10);    	
}

// Convert a byte array to a hex string 
static void toHex(int len, const char *src, char *dst)
{
    int i;
    for (i = 0; i < len; i++)
    {
        unsigned char ch = src[i];
        byte2hex(&dst[i * 2],ch);
    }
    dst[2*len]='\0';
}

#define BYTES 48

// Test vector
int main()
{
    const char *sk=(const char *)"c4eab05d357007c632f3dbb48489924d552b08fe0c353a0d4a1f00acda2c463afbea67c5e8d2877c5e3bc397a659949e";
    char prv[BYTES],pub[BYTES],sig[2*BYTES];
    char buff[256],m[2];
    int res;
    printf("private key= "); puts(sk); 
    fromHex(BYTES,sk,prv);
    ED376_KEY_PAIR(prv,pub);
    toHex(BYTES,pub,buff);
    printf("public key=  "); puts(buff);

    m[0]=0x03; // message to be signed
    ED376_SIGN(prv,pub,1,m,sig);
    toHex(2*BYTES,sig,buff);
    printf("signature=  "); puts(buff); 

    res=ED376_VERIFY(pub,1,m,sig);
    if (res)
        printf("Signature is valid\n");
    else
        printf("Signature is NOT valid\n");
}

File: test/tests376.rs
extern crate tlsecc;

use std::str;

use tlsecc::ed376::BYTES;
use tlsecc::ed376::KEY_PAIR;
use tlsecc::ed376::SIGN;
use tlsecc::ed376::VERIFY;


// Some utility functions for I/O and debugging

fn char2int(inp: u8) -> u8 {
    if inp>='0' as u8 && inp <='9' as u8 {
        return inp-'0' as u8;
    }
    if inp>='A' as u8 && inp <='F' as u8 {
        return inp-('A' as u8) +10;
    }
    if inp>='a' as u8 && inp <='f' as u8 {
        return inp-('a' as u8) +10;
    }
    return 0;
}

// string s better have even number of characters!
fn from_hex(ilen:usize,s: &str,x: &mut[u8]) {
    let mut pad:[u8;128]=[0;128];
    let c=s.as_bytes();
    let len=c.len();
    let mut lz=2*ilen-len;
    if 2*ilen<len {lz=0;}
    for i in 0..lz {
        pad[i]='0' as u8;
    }
    for i in lz..2*ilen {
        pad[i]=c[i-lz];
    }

    for i in 0..ilen {
        x[i]=char2int(pad[2*i])*16+char2int(pad[2*i+1]);
    }
}

fn printhex(len:usize,array: &[u8]) {
    for i in 0..len {
        print!("{:02X}", array[i])
    }
    println!("")
}

fn main() {
    const SK:&str="c4eab05d357007c632f3dbb48489924d552b08fe0c353a0d4a1f00acda2c463afbea67c5e8d2877c5e3bc397a659949e";

    let mut prv:[u8;BYTES]=[0;BYTES];
    let mut public:[u8;BYTES]=[0;BYTES];
    let mut sig:[u8;2*BYTES]=[0;2*BYTES];
    let mut m:[u8;2]=[0;2];

    from_hex(BYTES,&SK,&mut prv); 
    print!("private key= "); printhex(BYTES,&prv);
    KEY_PAIR(&prv,&mut public);
    print!("Public key= "); printhex(BYTES,&public);

    m[0]=0x03; // message to be signed
    SIGN(&prv,Some(&public),&m[0..1],&mut sig);
    print!("signature=  "); printhex(2*BYTES,&sig); 

    let res=VERIFY(&public,&m[0..1],&sig);
    if res {
        println!("Signature is valid");
    } else {
        println!("Signature is NOT valid");
    }
}